hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion6Strategy.java

```diff
@@ -31,34 +28,61 @@
  *     <li>48 bits - pseudorandom data to provide uniqueness.</li>
  * </ul>
  *
- * @apiNote This strategy is field-compatible with Version 1, with the time bits reordered for improved DB locality.
- *
  * @author Cedomir Igaly
+ * @apiNote This strategy is field-compatible with Version 1, with the time bits reordered for improved DB locality.
  */
 public class UuidVersion6Strategy implements UUIDGenerationStrategy, UuidValueGenerator {
 	public static final UuidVersion6Strategy INSTANCE = new UuidVersion6Strategy();
 
 	private static class Holder {
 		static final SecureRandom numberGenerator = new SecureRandom();
-		static final Instant EPOCH_1582 = LocalDate.of( 1582, 10, 15 )
+		static final long EPOCH_1582_SECONDS = LocalDate.of( 1582, 10, 15 )
 				.atStartOfDay( ZoneId.of( "UTC" ) )
-				.toInstant();
+				.toInstant().getEpochSecond();
+
 	}
 
-	private final Lock lock = new ReentrantLock( true );
-	private final AtomicLong clockSequence = new AtomicLong( 0 );
-	private long currentTimestamp;
+	private record State(long timestamp, int sequence) {
+		public State getNextState() {
+			final long now = instantToTimestamp();
+			if ( this.timestamp < now ) {
```

Comment:
Most of the time, `now` will be greater than `this.timestamp` i.e. the previous timestamp, so this code would always create a new random sequence.
This is what I was trying to say with [my comment](https://github.com/hibernate/hibernate-orm/pull/9020#discussion_r1778263870) about naming. If I understand the recommendations correctly, we have essentially two options:
* Always use a random clock sequence
* Use a random clock sequence only if `this.timestamp > now`

I don't know what this code is trying to attempt or what is a good strategy, but it makes sense to me to increase the clock sequence when two threads end up with the same timestamp to avoid collisions.

Reply:
`this.timestamp` is timestamp of the last state. In any "normal" situation current timestamp (`now`) should be either equal or larger than `timestamp`. Possible reasons for current timestamp being less that last one are mentioned in RFC - clock rollbacks or leap second handling. In that case generator will act as if they are equal and increment counter. Only in case if incrementing counter will make it to rollover, timestamp will be incremented and random counter generated.

About naming, `AtomicReference` holding last state is called `lastState`, and `timestamp`, and `counter` are just fields of `State` record. I am not sure that there is need to rename them, but it is not a problem if you think that this will make some big difference.

Reply:
My point is that a random clock sequence should only be generated if the previous timestamp is greater than the current timestamp. In all other cases, the clock sequence should just be incremented by 1.
Or if we decide to always use a random clock sequence, do so unconditionally.
The current code generates a random clock sequence if the current timestamp is greater than the previous timestamp, and increments the clock sequence if the current timestamp is smaller or equal to the previous timestamp. This is exactly opposite of what I understood should happen, which indicates to me that the naming you chose confused you.

Reply:
If you think that naming is problem, now I've renamed State object properties. The point is that timestamp at the moment of next state generation should normally be either equals to last one (then increment counter) or larger (generate random counter). In some exceptional situation when it is smaller, we simply assume that they are equal and incrementing counter. Only when counter is at its maximum value, timestamp is incremented and random counter is generated. Otherwise it will rollover to zero.